Update QPainter/QPicture mechanism used in drawing.

chris2oph opened the pull request:
The original drawing mechanism was generating lots of `QWarning` messages when running samples (e.g. htlbox, caret, etc.) and in some cases was not actually completely drawing every element of the sample. The issue was that the `QPicture` was being shared incorrectly between **window** and **dcclient** and attempts to start painting, update, etc. were generating console warnings.

chris2oph committed:
Ensured window and client use unique QPainters and ensure picture is drawn appropriately. Also changed a 'repaint' to an 'update' to follow QT widget painting best practise.

chris2oph committed:
Ensured that QPaiter member is deleted upon destruction of client DC.

MaartenBent commented on include/wx/qt/dcclient.h:
```diff
@@ -37,6 +39,8 @@ class WXDLLIMPEXP_CORE wxClientDCImpl : public wxWindowDCImpl
     wxDECLARE_CLASS(wxClientDCImpl);
     wxDECLARE_NO_COPY_CLASS(wxClientDCImpl);
+
+    QPicture* m_pic;
```
Maybe you can use [`wxScopedPtr<QPicture>`](https://docs.wxwidgets.org/3.1/classwx_scoped_ptr_3_01_t_01_4.html) instead of doing new/delete yourself?

chris2oph replied:
Thanks for the suggestion @MaartenBent however doing so would mean clients of the class would need to include `QScopedPointer`. They are currently completely insulated from QT and I believe it would be better to keep it this way.

vadz replied:
Note that Maarten wrote `wxScopedPtr`, not `QScopedPointer`.

vadz replied:
I still think the idea of using `wxScopedPtr` here is good. Also, I liked the old `pict` name while `pic` doesn't really associate with "picture" for me, could we please keep calling this `m_pict`?

And, as a final nitpick, could you please declare this variable above the `wxDECLARE_XXX` macros, which are supposed to be the very last thing in the class declaration?

vadz reviewed (commented):
Thanks for the fixes, we can merge this as is, but, if possible, it would be nice to address the various points below, especially the `repaint()` vs `update()` one. TIA!

chris2oph commented on src/qt/dcclient.cpp:
```diff
@@ -102,21 +108,27 @@ wxClientDCImpl::~wxClientDCImpl()
             {
                 // only force the update of the rect affected by the DC
-                widget->repaint( rect );
+                widget->update( rect );
```
No problem, I will extract this change into it's own pull request. I will ensure to add reasonable justification for this change when I do.

chris2oph committed:
Added a support class to handle setting of QT pictures. Replaced 'update' with 'repaint' on widget - this will be handled in a separate PR. Added a comment to improve understanding.

chris2oph committed:
Removed commented out calls to QtSetPicture().

chris2oph commented:
Hi @vadz , I have addressed the latest comments with my most recent commit. Let me know if there are any issues!

vadz commented on src/qt/dcclient.cpp:
```diff
@@ -88,10 +114,12 @@ wxClientDCImpl::~wxClientDCImpl()
 
         if ( m_window != NULL )
         {
+            QtPictureSetter(m_window, m_pic);
```
Sorry, but this can't work AFAICS as it creates and destroys the temporary object immediately, thus setting and unsetting the picture. The fix is trivial:
```suggestion
            QtPictureSetter pictureSetter(m_window, m_pic);
```
but what worries me is that this has somehow went unnoticed in testing, when it really shouldn't work at all.

vadz commented on src/qt/dcclient.cpp:
```diff
@@ -27,6 +27,28 @@
+
+    private:
+        wxWindow *m_window;
+    };
```
Should also add `wxDECLARE_NO_COPY_CLASS` for safety.

vadz commented on src/qt/dcclient.cpp:
```diff
@@ -27,6 +27,28 @@
 
+namespace
+{
+    class QtPictureSetter
```
Trivial, but could you please avoid indenting namespace contents? We don't do it elsewhere and consistency is nice.

vadz commented on src/qt/dcclient.cpp:
```diff
@@ -27,6 +27,28 @@
+        }
+
+    private:
+        wxWindow *m_window;
```
I'd make this const:
```suggestion
        wxWindow* const m_window;
```
and initialize it in the ctor initializer list then.

vadz commented on src/qt/dcclient.cpp:
```diff
@@ -88,10 +114,12 @@ wxClientDCImpl::~wxClientDCImpl()
         m_qtPainter->end();
         m_ok = false;
-        QPicture *pict = m_window->QtGetPicture();
+
```
```suggestion
```

vadz commented on src/qt/dcclient.cpp:
```diff
@@ -67,15 +89,19 @@ wxClientDCImpl::wxClientDCImpl( wxDC *owner )
 
-    QPicture *pic = win->QtGetPicture();
-    m_ok = m_qtPainter->begin( pic );
+    m_qtPainter = new QPainter();
```
It looks like this is going to be leaked. The code is rather confusing but AFAICS `m_qtPainter` is reset to NULL in the base class dtor, so the comment at the end of this class dtor is wrong. Isn't it?

chris2oph replied:
Thankfully the `m_qtPainter` member will not be leaked. The inheritance structure of the classes here is as follows:

`wxClientDCImpl` and `wxPaintDCImpl` both inherit from `wxWindowDCImpl`
`wxWindowDCImpl` inherits from `wxQtDCImpl`

The `m_qtPainter` member is declared in the base class (`wxQtDCImpl`), as well as being deleted in its destructor. Thanks to the deletion taking place in the base, we can be certain that the `QPainter` will be deleted whenever an instance of one of these classes goes out of scope.

vadz commented on src/qt/window.cpp:
```diff
@@ -1094,7 +1093,7 @@ bool wxWindowQt::QtHandlePaintEvent ( QWidget *handler, QPaintEvent *event )
             bool handled;
 
-            if ( m_qtPicture->isNull() )
+            if ( m_qtPicture == NULL)
```
```suggestion
            if ( m_qtPicture == NULL )
```

vadz reviewed (changes requested):
Sorry, still a few comments about this one, including a couple of rather important ones.

chris2oph committed:
Ensured there is actual usage of QtPictureSetter as well as using a scoped pointer to store the QPicture. Also rectified various minor issues.

chris2oph commented:
I have addressed your comments @vadz  with my latest commit, most of them were absolutely correct; thanks for the detailed review! The only comment I have not addressed is your suggestion of a memory leak, I have replied directly to the comment if you wish to see more.
